### ISSUE TITLE
on StochasticMux rates, scalloping loss contours

### ISSUE THREAD

**Issue by ejhumphrey** [pescadores/pescador#132]
I've just done something rather foolish (so this is documented user error), but wanted to share (a) the experience in case anyone else steps on it, and (b) an idea for improving upon StochasticMux to (potentially) help avoid this in the future.

First, I lost track of how many "unique" samples were in each bag, and the rate was too high. This can lead to "scalloped" training contours and _really_ slow learning ... here's a diagram for accuracy over iteration.

![image](https://user-images.githubusercontent.com/926379/37485429-cf401120-2861-11e8-9377-7bf301e0f71c.png)

What's happening is that the data stream isn't shuffling through the data fast enough, and overfitting a relatively small number of datapoints at a time. As the underlying muxes shuffle through the data, you get sharp drops in training accuracy when new streams are activated (more on this in a second). Then, because the model is sitting on the same data too long, it's not learning all that well, leading to a significant gap in performance with the validation data. 

tl;dr: turn your rate down.

Second, and perhaps more interesting, the sharp drops occur because the streams are all getting refreshed at roughly the same time. While the StochasticMux should sort this out over time (which it looks like is happening), it takes a while for this staggering to occur. One option would be to treat the first round of stream activations differently, instead drawing uniformly from, e.g. [1, rate]. I ... think .. this should be enough to decorrelate the various streamers from the outset.

I also *think* that change would mostly be confined to [mux.py:L754](https://github.com/pescadores/pescador/blob/master/pescador/mux.py#L754), with a little bit of logic in the __init__ method to indicate that it's the first round. 

thoughts?

**Comment by bmcfee** [comment]
> Second, and perhaps more interesting, the sharp drops occur because the streams are all getting refreshed at roughly the same time.

Do you still have the code from this example?  I'd be curious to know what the parameters `(rate, n_streamers, n_active)` were (as well as batch size for your training), and how that squares with the theoretical analysis we have for replacement rates.

**Comment by bmcfee** [comment]
>  One option would be to treat the first round of stream activations differently, instead drawing uniformly from, e.g. [1, rate]. I ... think .. this should be enough to decorrelate the various streamers from the outset.

That's an interesting idea -- essentially adding a random offset to each streamer slot's expected replacement time.  I don't think it should be default behavior, but it's a nice optimization.

**Comment by bmcfee** [comment]
Alternately, I wonder if using a stream cache could help smooth that out?
